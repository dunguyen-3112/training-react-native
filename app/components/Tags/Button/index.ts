import ButtonReadNow from './ButtonReadNow';
import CustomButton from './CustomButton';

export { ButtonReadNow, CustomButton };
