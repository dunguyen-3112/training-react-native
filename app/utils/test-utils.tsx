import { render, RenderOptions } from '@testing-library/react-native';
import React, { ReactElement } from 'react';
import { NavigationContainer } from '@react-navigation/native';

const AllTheProviders = ({ children }: { children: ReactElement }) => {
  return <NavigationContainer>{children}</NavigationContainer>;
};

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react-native';

export { customRender as render };
