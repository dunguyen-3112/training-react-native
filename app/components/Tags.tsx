import { StyleSheet, FlatList, Text, View, Alert } from 'react-native';
import React, { useCallback, useMemo, useState } from 'react';
import Tag, { TagProps } from './Tag';

const tags: TagProps[] = [
  {
    id: 1,
    name: 'Breakfast',
  },
  {
    id: 2,
    name: 'Dinner',
  },
  {
    id: 3,
    name: 'Dry fruits',
  },
  {
    id: 4,
    name: 'Fast Food',
  },
  {
    id: 5,
    name: 'Non-Veg',
  },
  {
    id: 6,
    name: 'Vegetables',
  },
  {
    id: 7,
    name: 'Lunch',
  },
  {
    id: 8,
    name: 'Greenish',
  },
  {
    id: 9,
    name: 'Fruits',
  },
];

const Tags = ({ marginTop }: { marginTop: number }) => {
  const [tag, setTag] = useState(tags[0].id);

  const handlePressTag = useCallback((id: number) => {
    setTag(id);
    Alert.alert(
      'Information',
      `ID: ${id}, Name: ${tags.find((tag) => tag.id === id)?.name}`
    );
  }, []);

  return (
    <View style={[styles.container, { marginTop: marginTop }]}>
      <FlatList
        data={tags}
        showsHorizontalScrollIndicator={false}
        horizontal
        ItemSeparatorComponent={() => <View style={styles.item} />}
        keyExtractor={(item) => item.id + ''}
        renderItem={({ item }: { item: TagProps }) => (
          <Tag {...item} onPress={handlePressTag} isActive={tag === item.id} />
        )}
      />
    </View>
  );
};

export default Tags;

const styles = StyleSheet.create({
  container: {
    height: 26,
  },
  item: {
    marginLeft: 7,
  },
});
