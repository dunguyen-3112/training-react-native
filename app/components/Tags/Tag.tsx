import { Alert, StyleSheet, TouchableOpacity } from 'react-native';
import React, { memo, useCallback } from 'react';
import { Categories, COLORS } from '@constants';
import { ICategory } from '@types';
import { Text } from '@components/common';

interface TagProps extends ICategory {
  isActive?: boolean;
  onPress?: (id: number) => void;
}

export { TagProps };

const Tag = ({ id, isActive, onPress }: TagProps) => {
  const handlePress = useCallback(() => {
    onPress && onPress(id);
  }, [id, onPress]);

  return (
    <Text
      onPress={handlePress}
      font={{ fontSize: 13 }}
      customStyle={styles.tag}
      {...(isActive && { color: COLORS.GREEN_COLOR })}
    >
      {Categories.find((c) => c.id === id)?.name}
    </Text>
  );
};

export default memo(Tag);

const styles = StyleSheet.create({
  tag: {
    backgroundColor: '#ECEDEC',
    borderRadius: 8,
    maxHeight: 26,
    paddingVertical: 4,
    paddingHorizontal: 10,
  },
  active: {
    backgroundColor: COLORS.GREEN_BLUR_COLOR,
    borderColor: COLORS.GREEN_COLOR,
    borderWidth: 1,
  },
});
