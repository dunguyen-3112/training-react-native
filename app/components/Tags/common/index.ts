import Tag, { TagProps } from './Tag';

export { Tag, TagProps };
