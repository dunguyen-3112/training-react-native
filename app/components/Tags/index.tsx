import { StyleSheet, FlatList, View } from 'react-native';
import React, { memo, useCallback, useState } from 'react';
import Tag, { TagProps } from './Tag';
import { Categories } from '@constants';

const Tags = ({
  marginTop,
  onSelect,
}: {
  marginTop: number;
  onSelect: (id: number) => void;
}) => {
  const [tag, setTag] = useState<number>();

  const handlePressTag = useCallback(
    (id: number) => {
      onSelect(id);
      setTag(id);
    },
    [onSelect]
  );

  return (
    <View style={[styles.container, { marginTop: marginTop }]}>
      <FlatList
        data={Categories}
        showsHorizontalScrollIndicator={false}
        horizontal
        ItemSeparatorComponent={() => <View style={styles.item} />}
        keyExtractor={(item, index) => index + ''}
        renderItem={({ item }: { item: TagProps }) => (
          <Tag {...item} onPress={handlePressTag} isActive={tag === item.id} />
        )}
      />
    </View>
  );
};

export default memo(Tags);

const styles = StyleSheet.create({
  container: {
    height: 26,
  },
  item: {
    marginLeft: 7,
  },
});
