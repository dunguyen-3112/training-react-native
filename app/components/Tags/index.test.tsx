import { render, screen, fireEvent } from '@utils/test-utils';
import { CATEGORIES } from '@constants';
import React from 'react';

import Tags, { ICategories } from './index';

const mockOnPress = jest.fn();

function setup(props: ICategories) {
  return render(<Tags {...props} />);
}

const component = setup({ categories: CATEGORIES });
describe('Categories', () => {
  it('should render List', () => {
    expect(component.toJSON()).toMatchSnapshot();

    CATEGORIES.forEach((category) => {
      expect(screen.getByText(category.name || '')).toBeTruthy();
    });
  });

  it('should render Item 1', () => {
    const { id, name } = CATEGORIES[1];

    const item = screen.getByText('inner');

    expect(item).toMatchSnapshot();
    if (item.parent?.parent) {
      fireEvent.press(item.parent.parent);
      expect(mockOnPress).toHaveBeenCalledWith(2);
    }
  });
});
