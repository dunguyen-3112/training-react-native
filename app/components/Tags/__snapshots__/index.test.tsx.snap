// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "height": 26,
      },
      Object {
        "marginTop": undefined,
      },
    ]
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      Array [
        Object {
          "id": 1,
          "name": "Breakfast",
        },
        Object {
          "id": 2,
          "name": "Dinner",
        },
        Object {
          "id": 3,
          "name": "Dryfruits",
        },
        Object {
          "id": 4,
          "name": "Lunch",
        },
        Object {
          "id": 5,
          "name": "Vegetables",
        },
        Object {
          "id": 6,
          "name": "Greenich",
        },
        Object {
          "id": 7,
          "name": "Non-Veg",
        },
        Object {
          "id": 8,
          "name": "Fast Food",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <Text
          onPress={[Function]}
          style={
            Array [
              Object {
                "borderColor": "#5C5C5C",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "textTransform": "none",
              },
              undefined,
              Object {
                "backgroundColor": "#ECEDEC",
                "borderRadius": 8,
                "maxHeight": 26,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          Breakfast
        </Text>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <Text
          onPress={[Function]}
          style={
            Array [
              Object {
                "borderColor": "#5C5C5C",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "textTransform": "none",
              },
              undefined,
              Object {
                "backgroundColor": "#ECEDEC",
                "borderRadius": 8,
                "maxHeight": 26,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          Dinner
        </Text>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <Text
          onPress={[Function]}
          style={
            Array [
              Object {
                "borderColor": "#5C5C5C",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "textTransform": "none",
              },
              undefined,
              Object {
                "backgroundColor": "#ECEDEC",
                "borderRadius": 8,
                "maxHeight": 26,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          Dryfruits
        </Text>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <Text
          onPress={[Function]}
          style={
            Array [
              Object {
                "borderColor": "#5C5C5C",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "textTransform": "none",
              },
              undefined,
              Object {
                "backgroundColor": "#ECEDEC",
                "borderRadius": 8,
                "maxHeight": 26,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          Lunch
        </Text>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <Text
          onPress={[Function]}
          style={
            Array [
              Object {
                "borderColor": "#5C5C5C",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "textTransform": "none",
              },
              undefined,
              Object {
                "backgroundColor": "#ECEDEC",
                "borderRadius": 8,
                "maxHeight": 26,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          Vegetables
        </Text>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <Text
          onPress={[Function]}
          style={
            Array [
              Object {
                "borderColor": "#5C5C5C",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "textTransform": "none",
              },
              undefined,
              Object {
                "backgroundColor": "#ECEDEC",
                "borderRadius": 8,
                "maxHeight": 26,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          Greenich
        </Text>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <Text
          onPress={[Function]}
          style={
            Array [
              Object {
                "borderColor": "#5C5C5C",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "textTransform": "none",
              },
              undefined,
              Object {
                "backgroundColor": "#ECEDEC",
                "borderRadius": 8,
                "maxHeight": 26,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          Non-Veg
        </Text>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <Text
          onPress={[Function]}
          style={
            Array [
              Object {
                "borderColor": "#5C5C5C",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "textTransform": "none",
              },
              undefined,
              Object {
                "backgroundColor": "#ECEDEC",
                "borderRadius": 8,
                "maxHeight": 26,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          Fast Food
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
